python/podio/test_CodeGen.py

```diff
@@ -31,3 +37,15 @@ def test_add(self):
         self.assertEqual(len(daughter_particle.parents()), 0)
         daughter_particle.addparents(parent_particle)
         self.assertEqual(len(daughter_particle.parents()), 1)
+
+
+class CollectionSubscriptTest(unittest.TestCase):
+    """Collection subscript test"""
+
+    def test_bound_check(self):
+        collection = nsp.EnergyInNamespaceCollection()
+        _ = collection.create()
+        self.assertEqual(len(collection), 1)
+        with self.assertRaises(IndexError):
+            _ = collection[20]
```

Comment:
This fails on `dev3` due to some overload resolution issue. I would assume it's a ROOT issue that we are hitting here, but I am not sure yet which one.

Reply:
Well an exception is thrown, but of a different type than expected:
```txt
  Traceback (most recent call last):
    File "/home/runner/work/podio/podio/python/podio/test_CodeGen.py", line 50, in test_bound_check
      _ = collection[20]
    File "/home/runner/work/podio/podio/python/podio/pythonizations/collection_subscript.py", line 22, in get_item
      return self.at(i)
  TypeError: none of the 2 overloaded methods succeeded. Full details:
    nsp::EnergyInNamespace nsp::EnergyInNamespaceCollection::at(size_t index) =>
      out_of_range: deque::_M_range_check: __n (which is 20)>= this->size() (which is 1)
    nsp::MutableEnergyInNamespace nsp::EnergyInNamespaceCollection::at(size_t index) =>
      out_of_range: deque::_M_range_check: __n (which is 20)>= this->size() (which is 1)
```

According to the [docs](https://cppyy.readthedocs.io/en/latest/functions.html#overloads-and-exceptions) if there is an exception then cppyy checks the rest of overloads and if they all throw the same type then the exception is propagated, if they all throw different types then `TypeError` is risen
Here both overloads throws `std::out_of_range` but somehow it isn't recognized as the same type

The possible solutions I could think of now are:
- explicitly select overload - here the overloads are const and non-const. Selection between const and non-const was added in cppyy-1.7.0, nightlies have 1.6.2
- change pythonization to catch either `cppyy.gbl.std.out_of_range` or `TypeError` and raise `IndexError`- I don't like it becasue it's also used for other things, like actual type mismatch, eg. calling with `str` instead of `int`
- allow both types of exception in test - getting an exception is already an improvement, the message says it's out of range. The users get confused why it's `TypeError` not `IndexError`

Reply:
Maybe this can be boiled down to a small reproducer that we can submit to ROOT because they probably didn't want to introduce too many "breaking" changes in their update of the cppyy they bundle.

Reply:
Opened an issue wlav/cppyy#230 as the problems appears also outside of ROOT bundle

Reply:
I also made the ROOT folks aware via https://github.com/root-project/root/issues/15375